rollup/cmd/proposer_tool/app/app.go

```diff
@@ -83,28 +92,33 @@ func action(ctx *cli.Context) error {
 		log.Crit("failed to connect l2 geth", "config file", cfgFile, "error", err)
 	}
 
-	genesisHeader, err := l2Client.HeaderByNumber(subCtx, big.NewInt(0))
+	startL2BlockHeight := ctx.Uint64(utils.StartL2BlockFlag.Name)
+	startL2Block, err := l2Client.BlockByNumber(context.Background(), big.NewInt(int64(startL2BlockHeight)))
 	if err != nil {
-		return fmt.Errorf("failed to retrieve L2 genesis header: %v", err)
+		log.Crit("failed to get start l2 block", "startL2BlockHeight", startL2BlockHeight, "error", err)
 	}
 
-	chunk := &encoding.Chunk{
-		Blocks: []*encoding.Block{{
-			Header:         genesisHeader,
-			Transactions:   nil,
-			WithdrawRoot:   common.Hash{},
-			RowConsumption: &gethTypes.RowConsumption{},
-		}},
-	}
+	chunk := &encoding.Chunk{Blocks: []*encoding.Block{{Header: startL2Block.Header()}}}
 
-	var dbChunk *orm.Chunk
-	dbChunk, err = orm.NewChunk(db).InsertChunk(subCtx, chunk, encoding.CodecV0, rutils.ChunkMetrics{})
+	prevChunk, err := orm.NewChunk(dbForReplay).GetParentChunkByBlockNumber(subCtx, startL2BlockHeight)
 	if err != nil {
-		log.Crit("failed to insert chunk", "error", err)
+		log.Crit("failed to get previous chunk", "error", err)
+	}
+
+	var startQueueIndex uint64
+	if prevChunk != nil {
+		startQueueIndex = prevChunk.TotalL1MessagesPoppedBefore + prevChunk.TotalL1MessagesPoppedInChunk
+	}
+
+	for _, tx := range startL2Block.Transactions() {
+		if tx.Type() == gethTypes.L1MessageTxType {
+			startQueueIndex++
```

Comment:
For this to work with CodecV7 from any `startL2Block` you need to also take the message queue hashes from `prevChunk`. You can then apply all msg queue hashes while summing up the L1 message count and set the final hash in the fake chunk you're creating.

Reply:
added a comment in this commit: [be3f189](https://github.com/scroll-tech/scroll/pull/1645/commits/be3f189715eabf54286be33fde3f2bf1e9833b14)